## Contents
    01.Importing libraries
    02.Importing Data
    03.Aggregating Data

# 01.Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02.Importing Data

In [2]:
path=r'C:\Users\rache\Documents\Instacart Basket Analysis'

In [3]:
#importing orders_products_merged_ver2.pkl
ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_ver2.pkl'))

# 03.Aggregating Data

### In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [4]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

#### The results from the entire dataframe include a full list of departments where the limited subset only had 8 departments. The means of the full dataframe are the true means, whereas means from the subset were based on a limited number of rows. For example, with the subset, the mean for dept 17 was 11.294069 but with the entire dataframe the mean for dept 17 is much higher at 15.694469.

### Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [5]:
#create a new column for maximum orders for each user
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [6]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  first_time_customer  product_id  add_to_cart_order  \
0                    NaN                 True         196                  1   
1                   15.0                False         196                  1   
2                   21.0                False         196                  1   
3                   29.0                False         196                  1   
4                   28.0                False         196                  1   

   reordered  ... Unnamed: 0  product_name aisle_id  department_id  prices  \
0          0  ...        192          Soda       77              7     9.0   
1          1  ...        192          Soda       77              7     9.0   
2          1  ...        192          Soda       77              7     9.0   
3          1  ...        192          Soda       77              7     9.0   
4          1  ...        192          Soda       77              7     9.0   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy        Average orders   
1  Mid-range product  Regularly busy    Slowest days        Average orders   
2  Mid-range product  Regularly busy    Slowest days           Most orders   
3  Mid-range product      Least busy    Slowest days        Average orders   
4  Mid-range product      Least busy    Slowest days           Most orders   

  max_order  
0        10  
1        10  
2        10  
3        10  
4        10  

[5 rows x 21 columns]

In [7]:
#deriving columns with loc function
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [8]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [9]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [10]:
#checking frequency
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [11]:
#checking flags assigned correctly
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

### The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). 

In [13]:
#grouping by loyalty flag and creating basic statistics of prices of products purchased by each type of customer
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

### Create a spending flag for each user based on the average price across all their orders using the following criteria: If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.” & If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [14]:
#create a new column for average prices for each user
ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [15]:
#deriving columns with loc function
ords_prods_merge.loc[ords_prods_merge['avg_price'] <10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['avg_price'] >=10, 'spending_flag'] = 'High spender'

In [16]:
#checking frequency
ords_prods_merge['spending_flag'].value_counts(dropna=False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [20]:
#checking flags assigned correctly
ords_prods_merge[['user_id', 'spending_flag', 'avg_price']].head(60)

user_id spending_flag  avg_price
0         1   Low spender   6.367797
1         1   Low spender   6.367797
2         1   Low spender   6.367797
3         1   Low spender   6.367797
4         1   Low spender   6.367797
5         1   Low spender   6.367797
6         1   Low spender   6.367797
7         1   Low spender   6.367797
8         1   Low spender   6.367797
9         1   Low spender   6.367797
10       15   Low spender   3.980556
11       15   Low spender   3.980556
12       15   Low spender   3.980556
13       15   Low spender   3.980556
14       15   Low spender   3.980556
15       19   Low spender   8.372059
16       19   Low spender   8.372059
17       19   Low spender   8.372059
18       21   Low spender   7.012683
19       31   Low spender   8.137793
20       31   Low spender   8.137793
21       43   Low spender   7.444898
22       43   Low spender   7.444898
23       52   Low spender   6.356287
24       52   Low spender   6.356287
25       52   Low spender   6.356287
26       52   Low spender   6.356287
27       52   Low spender   6.356287
28       52   Low spender   6.356287
29       52   Low spender   6.356287
30       52   Low spender   6.356287
31       52   Low spender   6.356287
32       52   Low spender   6.356287
33       52   Low spender   6.356287
34       52   Low spender   6.356287
35       52   Low spender   6.356287
36       52   Low spender   6.356287
37       67   Low spender   7.656790
38       67   Low spender   7.656790
39       67   Low spender   7.656790
40       67   Low spender   7.656790
41       67   Low spender   7.656790
42       67   Low spender   7.656790
43       67   Low spender   7.656790
44       67   Low spender   7.656790
45       67   Low spender   7.656790
46       67   Low spender   7.656790
47       67   Low spender   7.656790
48       67   Low spender   7.656790
49       67   Low spender   7.656790
50       67   Low spender   7.656790
51       67   Low spender   7.656790
52       67   Low spender   7.656790
53       67   Low spender   7.656790
54       67   Low spender   7.656790
55       67   Low spender   7.656790
56       81   Low spender   8.165385
57       81   Low spender   8.165385
58       82   Low spender   6.441221
59       82   Low spender   6.441221

###  Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows: If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”, If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”, & If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [17]:
#create a new column for order frequency for each user
ords_prods_merge['median_order_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [18]:
#deriving columns with loc function
ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] >20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_order_frequency'] >10) & (ords_prods_merge['median_order_frequency'] <=20), 'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] <=10, 'frequency_flag'] = 'Frequent customer'

In [19]:
#checking frequency
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: frequency_flag, dtype: int64

In [21]:
#checking flags assigned correctly
ords_prods_merge[['user_id', 'frequency_flag', 'median_order_frequency']].head(60)

user_id         frequency_flag  median_order_frequency
0         1  Non-frequent customer                    20.5
1         1  Non-frequent customer                    20.5
2         1  Non-frequent customer                    20.5
3         1  Non-frequent customer                    20.5
4         1  Non-frequent customer                    20.5
5         1  Non-frequent customer                    20.5
6         1  Non-frequent customer                    20.5
7         1  Non-frequent customer                    20.5
8         1  Non-frequent customer                    20.5
9         1  Non-frequent customer                    20.5
10       15      Frequent customer                    10.0
11       15      Frequent customer                    10.0
12       15      Frequent customer                    10.0
13       15      Frequent customer                    10.0
14       15      Frequent customer                    10.0
15       19      Frequent customer                     6.0
16       19      Frequent customer                     6.0
17       19      Frequent customer                     6.0
18       21      Frequent customer                     7.5
19       31      Frequent customer                     6.0
20       31      Frequent customer                     6.0
21       43      Frequent customer                    10.0
22       43      Frequent customer                    10.0
23       52      Frequent customer                     8.0
24       52      Frequent customer                     8.0
25       52      Frequent customer                     8.0
26       52      Frequent customer                     8.0
27       52      Frequent customer                     8.0
28       52      Frequent customer                     8.0
29       52      Frequent customer                     8.0
30       52      Frequent customer                     8.0
31       52      Frequent customer                     8.0
32       52      Frequent customer                     8.0
33       52      Frequent customer                     8.0
34       52      Frequent customer                     8.0
35       52      Frequent customer                     8.0
36       52      Frequent customer                     8.0
37       67      Frequent customer                     7.0
38       67      Frequent customer                     7.0
39       67      Frequent customer                     7.0
40       67      Frequent customer                     7.0
41       67      Frequent customer                     7.0
42       67      Frequent customer                     7.0
43       67      Frequent customer                     7.0
44       67      Frequent customer                     7.0
45       67      Frequent customer                     7.0
46       67      Frequent customer                     7.0
47       67      Frequent customer                     7.0
48       67      Frequent customer                     7.0
49       67      Frequent customer                     7.0
50       67      Frequent customer                     7.0
51       67      Frequent customer                     7.0
52       67      Frequent customer                     7.0
53       67      Frequent customer                     7.0
54       67      Frequent customer                     7.0
55       67      Frequent customer                     7.0
56       81  Non-frequent customer                    30.0
57       81  Non-frequent customer                    30.0
58       82       Regular customer                    11.5
59       82       Regular customer                    11.5

In [22]:
#exporting to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_flags.pkl'))